### DISCUSSION TITLE
How to properly persist WebView state at exit?

### DISCUSSION THREAD

**Discussion by CrendKing** [tauri-apps/tauri#4963] [Q&A] [answered]
I want to persist some state from the frontend right before the app exits. It doesn't seem to be as easy as I expect.

My initial attempt is simply create a `window.onbeforeunload` handler and let it call a command. It doesn't work because this is not an unload event. That event is usually triggered on refresh.

Second attempt is to emit an message to frontend inside `WindowEvent::CloseRequested` handler, block until for the Javascript to call my command, unblock and let the app exit. It doesn't work (I think) because sleeping in the event handler causes the whole event loop to stop, thus my command will never be triggered.

I got it work by preventing the `WindowEvent::CloseRequested` from closing by `api.prevent_close()`, emit the message, and finally `app_handle.exit()` inside the command. What's wrong is that I always get an error message in the console:
```
[0816/225145.587:ERROR:window_impl.cc(114)] Failed to unregister class Chrome_WidgetWin_0. Error = 0
```
I assume we shouldn't call `app_handle.exit()` inside a command?

Is there a more standard way to do this?

**Comment by abisar** [comment, marked as answer]
If you're handling the state from javascript and it has nothing to do with rust, simply handle it only on javasscript side:
https://tauri.app/v1/api/js/classes/window.webviewwindow/#oncloserequested

> What's wrong is that I always get an error message in the console:

https://bugs.chromium.org/p/chromium/issues/detail?id=113008
I wouldn't worry about it.
